External/FEXCore/Source/Interface/HLE/Thunks/Thunks.cpp

```diff
@@ -27,19 +30,74 @@ tags: glue|thunks
 
 struct LoadlibArgs {
     const char *Name;
-    uintptr_t CallbackThunks;
 };
 
 static thread_local FEXCore::Core::InternalThreadState *Thread;
 
+static __attribute__((aligned(16), naked, section("HostToGuestTrampolineTemplate"))) void HostToGuestTrampolineTemplate() {
+#if defined(_M_X86_64)
+  asm(
+    "lea 0f(%rip), %r11 \n"
+    "jmpq *0f(%rip) \n"
+    ".align 8 \n"
+    "0: \n"
+    ".quad 0, 0, 0, 0 \n" // TrampolineInstanceInfo
+  );
+#elif defined(_M_ARM_64)
+  asm(
+    "adr x11, 0f \n"
+    "ldr x16, [x11] \n"
+    "br x16 \n"
+    // Manually align to the next 8-byte boundary
+    // NOTE: GCC over-aligns to a full page when using .align directives on ARM (last tested on GCC 11.2)
+    "mov x0, x0 \n"
+    "0: \n"
+    ".quad 0, 0, 0, 0 \n" // TrampolineInstanceInfo
+  );
+#else
+#error Unsupported host architecture
+#endif
+}
+
+extern char __start_HostToGuestTrampolineTemplate[];
+extern char __stop_HostToGuestTrampolineTemplate[];
 
 namespace FEXCore {
     struct ExportEntry { uint8_t *sha256; ThunkedFunction* Fn; };
 
+    struct TrampolineInstanceInfo {
+      uintptr_t HostPacker;
+      uintptr_t CallCallback;
+      uintptr_t GuestUnpacker;
+      uintptr_t GuestTarget;
+    };
+
+    struct GuestcallInfo {
+      uintptr_t GuestUnpacker;
+      uintptr_t GuestTarget;
+
+      bool operator==(const GuestcallInfo&) const noexcept = default;
+    };
+
+    struct GuestcallInfoHash {
+      size_t operator()(const GuestcallInfo& x) const noexcept {
+        // Hash only the target address, which is generally unique.
+        // For the unlikely case of a hash collision, std::unordered_map still picks the correct bucket entry.
+        return std::hash<uintptr_t>{}(x.GuestTarget);
+      }
+    };
+
+    // Hash the least significant bits only
+    struct TruncatingSHA256Hash {
+      size_t operator()(const FEXCore::IR::SHA256Sum& SHA256Sum) const noexcept {
+        return std::hash<size_t>{}((const size_t&)SHA256Sum);
```

Comment:
As `SHA256Sum`'s has random bit distribution, we don't really need to re-hash here

Reply:
uniform bit distribution is probably better for the comment to explain why not needing to hash again isn't necessary.